### PR TITLE
Migrate Dropdown to Bootstrap 4.1.1

### PR DESCRIPTION
**Pull request by yamgent**
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [X] Other, please explain: Migration

Part of MarkBind/markbind#298.

**What changes did you make? (Give an overview)**
* Migrate all CSS classes from Bootstrap 3 to Bootstrap 4
* Add some hack to customized 'dropdown menu' slot, because the toggling of the dropdown menu now modification of classes on the inner list `<ul>`, rather than on the outside element `<li>/<div>`.
* Removed the caret.
* Fixed a Vue migration porting issue (forgot to port `this._slotContents` over).

**BREAKING CHANGES**

* All dropdown items must include "dropdown-item" class.
* All divider must change to the "dropdown-divider" class.

Old:
```html
<li><a href="#">Some random item</a></li>
<li role="separator" class="divider"></li>
<li><a href="#">Some more random item</a></li>
```

New:
```html
<li><a href="#" class="dropdown-item">Some random item</a></li>
<li role="separator" class="dropdown-divider"></li>
<li><a href="#" class="dropdown-item">Some more random item</a></li>
```

**Testing instructions:**

1. Build vue-strap in this branch using `npm run build`.
2. Copy `vue-strap.min.js` from the `dist` folder in this repository, to the `asset/js/` folder in `markbind` repository (NOTE: Activate the `bootstrap-migration` branch on `markbind` repository!)
3. Do `markbind init` on a new empty folder.
4. On its `index.md`, add this content (note: this is adapted from website-base):

```html
<menu>
  <dropdown text="Dropdown Normal">
    <li><a href="#" class="dropdown-item">Normal 1</a></li>
    <li><a href="#" class="dropdown-item">Normal 2</a></li>
  <li role="separator" class="dropdown-divider"></li>
    <li><a href="#" class="dropdown-item">Normal 3</a></li>
  </dropdown>

  <dropdown text="Dropdown Disabled" disabled>
    <li><a href="#" class="dropdown-item">Disabled 1</a></li>
    <li><a href="#" class="dropdown-item">Disabled 2</a></li>
  </dropdown>

  <dropdown text="Dropdown with Type" type="danger">
    <li><a href="#" class="dropdown-item">Danger 1</a></li>
    <li><a href="#" class="dropdown-item">Danger 2</a></li>
  </dropdown>

  <dropdown text="Dropdown Show" show>
    <li><a href="#" class="dropdown-item">Automatically shown 1</a></li>
    <li><a href="#" class="dropdown-item">Automatically shown 2</a></li>
  </dropdown>

  <dropdown text="Dropdown Slots">
    <ul slot="dropdown-menu" class="dropdown-menu">
      <li><a href="#" class="dropdown-item">Automatically shown 1</a></li>
      <li><a href="#" class="dropdown-item">Automatically shown 2</a></li>
    </ul>
  </dropdown>

  <dropdown>
    <li><a href="#" class="dropdown-item">No text 1</a></li>
    <li><a href="#" class="dropdown-item">No text 2</a></li>
  </dropdown>

  <dropdown text="No Items Dropdown">
  </dropdown>

  <dropdown type="info">
    <button slot="before" type="button" class="btn btn-info">Segmented</button>
    <li><a href="#dropdown" class="dropdown-item">Action</a></li>
  </dropdown>

  <dropdown text="Dropdown Slots with show" show>
    <ul slot="dropdown-menu" class="dropdown-menu">
      <li><a href="#" class="dropdown-item">Shown automatically 1</a></li>
      <li><a href="#" class="dropdown-item">Shown automatically 2</a></li>
    </ul>
  </dropdown>
</menu>
```

5. Run `markbind serve` and ensure that the appearances are not broken & the dropdown logic works (i.e. can open and close).

**Screenshots**

Old | New
-- | --
 |

### CODE REVIEW

File: src/Dropdown.vue [hunk 1]
```diff
@@ -1,27 +1,25 @@
...
-    <slot name="dropdown-menu">
-      <ul class="dropdown-menu">
+    <slot name="dropdown-menu" :class="menuClasses">
+      <ul class="dropdown-menu" :class="menuClasses">
         <slot></slot>
       </ul>
     </slot>
   </li>
   <div v-else ref="dropdown" :class="classes">
     <slot name="before"></slot>
```
Comment (Chng-Zhi-Xuan) [inline src/Dropdown.vue]: Currently the `.btn` class has `padding: .375rem .75rem;`, which makes Segmented buttons quite fat. maybe add some conditional rendering for the dropdown button to slim down when it has contents within the **before slot**?

File: src/Dropdown.vue [hunk 2]
```diff
@@ -48,15 +46,22 @@ export default {
     },
     type: {
       type: String,
-      default: 'default'
+      default: ''
```
Comment (Chng-Zhi-Xuan) [inline src/Dropdown.vue]: We should choose one of the official button types from Bootstrap 4 (probably "light") and make it the default. Any specific styling (such as border), we should override below in the Vue component.

File: src/Dropdown.vue [hunk 3]
```diff
@@ -73,34 +78,48 @@ export default {
       return this.$parent && (this.$parent.menu || this.$parent.submenu)
     },
     slots () {
-      return this._slotContents
+      return this.$slots.content
```
Comment (Chng-Zhi-Xuan) [inline src/Dropdown.vue]: Shouldn't it be `this.$slots.default`?, I don't see a `<slot name="content"></slot>` anywhere

File: src/Dropdown.vue [hunk 4]
```diff
@@ -73,34 +78,48 @@ export default {
...
+    },
+    hideDropdownMenu() {
+      this.show = false;
+      const $el = $(this.$refs.dropdown);
+      $el.findChildren('ul').each(ul => ul.classList.remove('show'));
+    },
+    showDropdownMenu() {
+      this.show = true;
+      const $el = $(this.$refs.dropdown);
+      $el.findChildren('ul').each(ul => ul.classList.add('show'));
```
Comment (Chng-Zhi-Xuan) [inline src/Dropdown.vue]: - `this.$refs.dropdown` returns a HTML Element which can be manipulated directly, is there a reason to have `const $el`?
- Also maybe use `.toggle(class, true|false)`?

e.g.
`this.$refs.dropdown.findChildren('ul').each(ul => ul.classList.toggle('show', false));`

**Comment by Chng-Zhi-Xuan** [comment]
**Bug found**
- `show` doesn't work with dropdown menus inserted with slot

```html
<dropdown text="Dropdown Slots" show>
  <ul slot="dropdown-menu" class="dropdown-menu">
    <li><a href="#" class="dropdown-item">NOT Automatically shown 1</a></li>
    <li><a href="#" class="dropdown-item">NOT Automatically shown 2</a></li>
  </ul>
</dropdown>
```

File: src/Dropdown.vue (hunk 2 above)
Comment (Chng-Zhi-Xuan) [reply src/Dropdown.vue]: Continuation of discussion: We agree to override `.btn-light` class in `markbind.css`, as a separate PR.

**Comment by yamgent** [comment]
* Changes requested made.
* Fix the bug mentioned. I updated the PR post to include "a dropdown menu inserted with slot" to verify that it works.

File: src/Dropdown.vue [hunk 5]
```diff
@@ -142,4 +148,8 @@ export default {
   padding: 0;
   border: 0;
 }
+
+.btn-with-before {
+  padding-left: 0;
```
Comment (Chng-Zhi-Xuan) [inline src/Dropdown.vue]:
```css
padding-left: 0.2rem;
padding-right: 0.4rem;
```
This looks closer to the old style.

File: src/Dropdown.vue [hunk 6]
```diff
@@ -142,4 +148,8 @@ export default {
...
+}
 </style>
```
Comment (Chng-Zhi-Xuan) [inline src/Dropdown.vue]: newline at EOF

File: src/Dropdown.vue [hunk 7]
```diff
@@ -46,12 +46,16 @@ export default {
...
+      default: 'light'
     }
   },
   computed: {
     btnType () {
-      return `btn-${this.type}`;
+      if (this.type !== '') {
+        return `btn-${this.type}`;
+      } else {
+        return '';
```
Comment (Chng-Zhi-Xuan) [inline src/Dropdown.vue]: Empty String won't appear in the logic as the default value is 'light', unless the user explicitly passes in an empty string (unlikely). You should refactor to account for this.

File: src/Dropdown.vue [hunk 8]
```diff
@@ -58,7 +58,7 @@ export default {
       }
     },
     classes () {
-      return [{disabled: this.disabledBool}, this.class, this.isLi ? 'dropdown' : this.inInput ? 'input-group-btn': 'btn-group']
+      return [{disabled: this.disabledBool}, this.class, this.isLi ? 'dropdown' : this.inInput ? 'input-group-append': 'btn-group']
```
Comment (Chng-Zhi-Xuan) [inline src/Dropdown.vue]: `Input` component not migrated. Should trim some logic here.

File: src/Dropdown.vue [hunk 9]
```diff
@@ -48,15 +46,22 @@ export default {
...
-      return `btn-${this.type}`;
+      if (this.type !== '') {
+        return `btn-${this.type}`;
+      } else {
+        return '';
+      }
     },
     classes () {
-      return [{open: this.showBool, disabled: this.disabledBool}, this.class, this.isLi ? 'dropdown' : this.inInput ? 'input-group-btn': 'btn-group']
+      return [{disabled: this.disabledBool}, this.class, this.isLi ? 'dropdown' : this.inInput ? 'input-group-append': 'btn-group']
```
Comment (Chng-Zhi-Xuan) [inline src/Dropdown.vue]: `this.inInput` is part of the **Input Component** which was not migrated, we can trim off some logic here.

**Comment by yamgent** [comment]
* Changes requested made.

**Pull request merged**